olive/data/component/pre_process_data.py

```diff
@@ -122,7 +122,119 @@ def _tokenizer_and_align_labels(examples):
         tokenized_inputs["label"] = new_labels
         return tokenized_inputs
 
-    tokenized_datasets = _huggingface_pre_precess_helper(
+    tokenized_datasets = _huggingface_pre_process_helper(
         _dataset, model_name, input_cols, label_cols, _tokenizer_and_align_labels, **kwargs
     )
     return BaseDataset(tokenized_datasets, label_cols=["label"], max_samples=max_samples)
+
+
+@Registry.register_pre_process()
+def text_generation_huggingface_pre_process(
+    _dataset, model_name, input_cols, seqlen, stride=None, max_samples=None, **kwargs
+):
+    """Pre-process data.
+
+    Args:
+        _dataset (object): Data to be pre-processed.
+        model_name (str): Name of the huggingface model.
+        input_cols (list): List of input columns.
+        seqlen (int): Length of the sequence.
+        stride (int): Stride to use when splitting the sequence. I
+            If None, the sequence is split into non-overlapping sequences. No context is used.
+            When stride is not None, we use a sliding window approach to split the sequence. The stride is
+            also used as context length.
+        max_samples (int): Maximum number of samples to use.
+        **kwargs: Additional arguments.
+            joiner (str): Delimiter to use when joining the rows of the input columns.
+            random_seed (int): Random seed to use. If not None, we use the random seed to choose the starting
+                point of the sequence.
+
+    Returns:
+        object: Pre-processed data.
+    """
+    from random import Random
+
+    from datasets import Dataset as HFDataset
+    from transformers import AutoTokenizer
+
+    # get tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+    # gather text from all input columns
+    text_list = []
+    for input_col in input_cols:
+        text_list += _dataset[input_col]
+    # delimiter between the text sequences
+    joiner = kwargs.get("joiner", " ")
+    text = joiner.join(text_list)
+
+    # in order to make processing faster we will only tokenize as much as needed
+    # assumes that num words > num tokens
+    split_text = text.split(" ")
+    num_text = len(split_text)
+
+    tokenized_inputs = {
+        "input_ids": [],
+        "target_ids": [],
+    }
+
+    random_seed = kwargs.get("random_seed")
+    if random_seed is None:
+        # no randomization, just use contiguous blocks of tokens
+        max_text = max_samples * seqlen if max_samples is not None else num_text
+        max_text = min(max_text, num_text)
```

Comment:
Different models have different tokenizers(encoders), like BPE, wordPiece, Unigram and etc. which may generate various sequence lengths.

seq_len means the tokenized sequence length at the most time from I know.

It is better to compare the length at the same level. **Not the word piece vs token**

https://huggingface.co/docs/transformers/tokenizer_summary

Reply:
We are using `seqlen` at the token level. 
This part is used as a heuristic to speed up tokenization. Most of the time (maybe everytime), the number of tokens is more than the number of words. So, we are taken a chunk of text with the `max_text` length to only tokenize as much as we need. Otherwise, it takes a long time to tokenize all of the text when we are not going to use them all. 

You can see that `seqlen` of the encodings is used to get the actual inputs and targets.

Reply:
Also, from my understanding, 

Different sentence may not be relevant with each other. In this [example](https://huggingface.co/docs/transformers/perplexity), they did with the same way to concatenate.
But in some cases, it is more like to calculate PPL per sentence then to get the mean-PPL.
![image](https://github.com/microsoft/Olive/assets/13343117/8e41ac2b-6471-4309-a436-63a28223bc6a)

I do not know which should be the best way to reproduce some model's benchmark report. 
Could you help validation that with a real model?

Reply:
I was thinking about this case too. I have seen some examples where each sentence is padded or truncated and made into a sample. 
I am thinking of maybe introducing different modes. Something like:
- joint-block (join all sentences. No sliding window)
- joint-sliding-window (join all sentences, sliding window + context)
- joint-random (join all sentences, random start point)
- individual-block (each sentence is a sample. truncated or padded to max length)
- individual-random (each sentence is a sample. randomly choose sentences. truncated or padded to max length) 

I think that would cover most of the ways of measuring perplexity. 

Can I do it in a follow up PR? I need this dataset and ppl support for sparsity work item. and currently don't have time to implement all modes and validate with models.

Reply:
For sure.